src/flint/types/fmpq_mpoly.pyx

```diff
@@ -0,0 +1,982 @@
+from flint.flint_base.flint_base cimport flint_mpoly
+from flint.flint_base.flint_base cimport flint_mpoly_context
+
+from flint.utils.typecheck cimport typecheck
+from flint.utils.flint_exceptions import IncompatibleContextError
+
+from flint.types.fmpq cimport fmpq, any_as_fmpq
+from flint.types.fmpz_vec cimport fmpz_vec
+from flint.types.fmpq_vec cimport fmpq_vec
+
+from flint.types.fmpz cimport fmpz, any_as_fmpz
+from flint.types.fmpz_mpoly cimport fmpz_mpoly
+
+from flint.flintlib.flint cimport *
+from flint.flintlib.fmpq cimport fmpq_set, fmpq_one
+from flint.flintlib.fmpq_mpoly cimport *
+from flint.flintlib.fmpq_mpoly_factor cimport *
+
+from flint.flintlib.fmpz cimport fmpz_init_set
+from flint.flintlib.fmpz_mpoly cimport fmpz_mpoly_set
+
+from cpython.object cimport Py_EQ, Py_NE
+cimport cython
+cimport libc.stdlib
+
+cdef dict _fmpq_mpoly_ctx_cache = {}
+
+
+@cython.auto_pickle(False)
+cdef class fmpq_mpoly_ctx(flint_mpoly_context):
+    """
+    A class for storing the polynomial context
+
+    :param nvars: The number of variables in the ring
+    :param ordering:  The term order for the ring
+    :param names:  A tuple containing the names of the variables of the ring.
+
+    Do not construct one of these directly, use `fmpz_mpoly_ctx.get_context`.
+    """
+
+    _ctx_cache = _fmpq_mpoly_ctx_cache
+
+    def __init__(self, slong nvars, ordering, names):
+        if ordering == "lex":
+            fmpq_mpoly_ctx_init(self.val, nvars, ordering_t.ORD_LEX)
+        elif ordering == "deglex":
+            fmpq_mpoly_ctx_init(self.val, nvars, ordering_t.ORD_DEGLEX)
+        elif ordering == "degrevlex":
+            fmpq_mpoly_ctx_init(self.val, nvars, ordering_t.ORD_DEGREVLEX)
```

Comment:
We should have a Python enum or at least a dict or something to represent this mapping.